### ISSUE TITLE
MVP for the GRIN Project

### ISSUE THREAD

**Issue by GunpowderGuy** [grin-compiler/grin#132]
The GRIN project has been dormant recently, and I’d like to reignite interest by developing an MVP (Minimum Viable Product) that includes essential features for a functional backend capable of handling real-world programs. I believe the Idris2 backend could serve this purpose with a few key enhancements: basic FFI (Foreign Function Interface) support and garbage collection.
Proposed Features:

    Foreign Function Interface (FFI)
        Currently, only GRIN backends can define FFI functions (RTS functions). To quickly enable user programs to perform FFI, a potential solution is to implement an RTS function that indirectly calls the user-specified external function. While this is a "quick and dirty" approach, it could expedite FFI support for user programs in the MVP phase.

    Garbage Collection
        I’ve found little concrete documentation about using LLVM’s GC facilities in an actual compiler. While LLVM's GC infrastructure is powerful, it might be too complex for an MVP. Although reference counting is simple to implement, it would likely result in poor performance without optimizations.
        A potential compromise could be integrating a plug-and-play garbage collector like Boehm. It offers a balance between ease of integration and reasonable performance, allowing us to move forward quickly without significant performance trade-offs.

Call for Feedback and Partners:

I’m looking for feedback on this approach and seeking partners who would like to work this out alongside me. I will be available for full production starting in mid-December, and I’m eager to collaborate to push GRIN forward with a solid foundation for real-world usage.

**Comment by luc-tielen** [comment]
I haven't spoken with @csabahruska in a while, but last I heard there was a lot of focus on tooling (like GHC-WPC) to get better insights while working on GRIN.

**Comment by csabahruska** [comment]
Recently I've learned GHC Cmm language, and it is also suitable to use as a GRIN target language. In that case we also could use all GHC RTS features, like GC, STM, concurrency. With GHC's LLVM backend which translates Cmm to LLVM we can have the benefits of LLVM also. IMO this would be the simplest way to have a GRIN MVP.
What do you think?

**Comment by luc-tielen** [comment]
@GunpowderGuy you could consider using https://GitHub.com/luc-tielen/llvm-codegen. Its not on hackage (I should add it there 😅) and has support for newer llvm versions, with a similar API. But it can only generate IR, not parse it. (Idk if grin needs that capability.)

**Comment by GunpowderGuy** [comment]
@luc-tielen I dont think Grin needs to parse llvm. 

Would using your library be a closer to rewriting the code gen or be a more plug and play change? If the latter, i would personally prefer to write a new code gen ( targetting llvm or otherwise ) in idris2.

**Comment by luc-tielen** [comment]
@GunpowderGuy the API is _very_ similar, with a few small changes for more type safety (but I think that was mostly internal).

**Comment by GunpowderGuy** [comment]
Kickass, then i could try to port GRIN to use that library. Btw here is the current code : https://github.com/csl-club/Idris2-Grin

That repo contains both the idris2 backend and a GRIN version modified by Z-snails . No other Grin brach/fork has more recent changes.

@csabahruska Would you still be interested in writing a new code gen in the near term? or should i go ahead updating the current one ?

**Comment by GunpowderGuy** [comment]
@luc-tielen Could you upload your llvm library to stackage?

**Comment by GunpowderGuy** [comment]
@csabahruska I am attempting to upgrade Grin's dependencies, but the package "infix functor" doesnt compile with modern stackage versions and hasnt been updated since 2017

**Comment by csabahruska** [comment]
I've forked and updated the `functor-infix` to compile with newer GHC and TH. Please use this fork for now.
https://github.com/csabahruska/functor-infix

**Comment by GunpowderGuy** [comment]
Thanks csaba. I have kept updating the dependencies and found out there is a current llvm hs fork that is trying to get merged . 

https://github.com/llvm-hs/llvm-hs/pull/430

We should be able to use it instead of revamping the backend. However a bug in that branch is currently holding me back from completing GRIN's transition to a modern stackage version. I hope it gets solved in the following days

**Comment by GunpowderGuy** [comment]
@luc-tielen Given you seem to be an expert on llvm, could i ask you some questions privately about it? The point is learn how to best use it, or whether i should focus on another tecnology ( such as cranelift )

**Comment by luc-tielen** [comment]
Yeah sure, but just keep in mind that I am super busy irl so finding a time to talk may not be easy. 😅

**Comment by stevanmilic** [comment]
Hi @csabahruska @GunpowderGuy, for some time I’ve been working on my own programming language: https://github.com/fuselang/fuse (site: https://fuselang.org/). Which successfully uses grin as compiler back-end and   includes it into the fuse toolchain.

I came up with a working (MVP) grin build & executable using llvm-15: https://github.com/fuselang/grin/commit/d2a880ea03e136b4252947e95493cfef22b36b60.

As a follow-up I’ve included an option to use boehm-gc garbage collector in grin: https://github.com/fuselang/grin/commit/99fb24f0cc2492394f058cc46fbc3514c58c62ee

Let me know if you would be interested in merging this into the main grin compiler and referencing the fuse implementation. 

P.S. @csabahruska and I had a lengthy call a couple of years ago on using grin in fuse language, with some of his great feedback and ideas :)
